File: src/blockchain.py
import requests
import time
import logging
import os
import sys
from dotenv import load_dotenv

# Add the parent directory to the path to make imports work from any directory
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# Now we can import our modules
try:
    # Try importing with the src prefix first (when running as a module)
    from src.database import get_all_contracts, update_blockchain_metrics
except ModuleNotFoundError:
    # Fall back to direct imports (when running the script directly)
    from database import get_all_contracts, update_blockchain_metrics

# Load environment variables
load_dotenv()

# Set up logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

BSCSCAN_API_KEY = os.getenv("BSCSCAN_API_KEY")
BSCSCAN_BASE_URL = "https://api.bscscan.com/api"
REQUEST_DELAY = 0.25  # 250ms delay between requests to avoid rate limits

def fetch_holders_count(contract_address):
    """
    Fetch the number of token holders for a contract.

    Args:
        contract_address (str): The contract address to query.

    Returns:
        int: The number of token holders, or 0 if the request fails.
    """
    try:
        # Unfortunately, BscScan API doesn't provide a direct endpoint for holder count
        # We'll need to fetch the token info and parse it, or use alternative approaches

        # For demo purposes, we'll use a workaround: getting a sample of holders and estimating
        params = {
            'module': 'token',
            'action': 'tokenholderlist',
            'contractaddress': contract_address,
            'page': 1,
            'offset': 10,  # Request only a small number to stay within rate limits
            'apikey': BSCSCAN_API_KEY
        }

        logger.info(f"Fetching holders for contract: {contract_address[:8]}...")
        response = requests.get(BSCSCAN_BASE_URL, params=params, timeout=10)
        data = response.json()

        if data['status'] == '1':
            # If successful, return the total count if provided, or just the count of returned holders
            if 'countTokenHolders' in data:
                holders_count = int(data['countTokenHolders'])
            else:
                # Otherwise, just count the returned holders (less accurate)
                holders_count = len(data['result'])

            logger.info(f"Contract {contract_address[:8]} has {holders_count} holders")
            return holders_count
        else:
            logger.warning(f"Failed to get holders for {contract_address[:8]}: {data.get('message', 'Unknown error')}")
            return 0

    except Exception as e:
        logger.error(f"Error fetching holders for {contract_address[:8]}: {e}")
        return 0

def fetch_transfers_24h(contract_address):
    """
    Fetch and count token transfers in the last 24 hours.

    Args:
        contract_address (str): The contract address to query.

    Returns:
        int: The number of transfers in the last 24 hours, or 0 if the request fails.
    """
    try:
        # Calculate timestamp for 24 hours ago
        timestamp_24h_ago = int(time.time()) - 86400

        params = {
            'module': 'account',
            'action': 'tokentx',
            'contractaddress': contract_address,
            'startblock': 0,
            'endblock': 99999999,
            'sort': 'desc',
            'apikey': BSCSCAN_API_KEY
        }

        logger.info(f"Fetching recent transfers for contract: {contract_address[:8]}...")
        response = requests.get(BSCSCAN_BASE_URL, params=params, timeout=10)
        data = response.json()

        if data['status'] == '1':
            # Count transfers in the last 24 hours
            transfers_24h = sum(1 for tx in data['result'] if int(tx['timeStamp']) > timestamp_24h_ago)
            logger.info(f"Contract {contract_address[:8]} has {transfers_24h} transfers in the last 24 hours")
            return transfers_24h
        else:
            logger.warning(f"Failed to get transfers for {contract_address[:8]}: {data.get('message', 'Unknown error')}")
            return 0

    except Exception as e:
        logger.error(f"Error fetching transfers for {contract_address[:8]}: {e}")
        return 0

def update_blockchain_data():
    """Update blockchain data for all contracts in the database."""
    logger.info("Starting blockchain data update...")

    contracts = get_all_contracts()

    if not contracts:
        logger.warning("No contracts found in the database. Skipping blockchain data update.")
        return

    logger.info(f"Found {len(contracts)} contracts to update")

    for contract in contracts:
        try:
            # Fetch data from BscScan
            holders = fetch_holders_count(contract)
            time.sleep(REQUEST_DELAY)  # Avoid rate limits

            transfers = fetch_transfers_24h(contract)
            time.sleep(REQUEST_DELAY)  # Avoid rate limits

            # Update database
            current_timestamp = int(time.time())
            update_blockchain_metrics(contract, holders, transfers, current_timestamp)

            logger.info(f"Updated metrics for {contract[:8]}: {holders} holders, {transfers} recent transfers")

        except Exception as e:
            logger.error(f"Error updating blockchain data for {contract[:8]}: {e}")
            continue

    logger.info("Blockchain data update completed")

if __name__ == "__main__":
    update_blockchain_data()

File: src/blockchain.js
const fetch = require('node-fetch');
const Database = require('./database');

class Blockchain {
  // Configuration
  static etherscan_api_key = process.env.ETHERSCAN_API_KEY || '';

  static async update_blockchain_data() {
    console.log('Starting blockchain data update...');
    const updated_coins = [];

    try {
      // Get all coins from the database
      const coins = await Database.get_all_coins();

      if (!coins || coins.length === 0) {
        console.log('No coins found in database to update');
        return updated_coins;
      }

      console.log(`Found ${coins.length} coins in database to update`);

      // Process each coin - fetch blockchain data
      for (const coin of coins) {
        try {
          // Skip coins without contract addresses
          if (!coin.contract) {
            console.warn(`Skipping coin ${coin.name} (${coin.symbol}) - no contract address`);
            continue;
          }

          // Fetch holder count from Etherscan (example)
          const holders = await this.get_token_holder_count(coin.contract);

          // Fetch transfer count for last 24 hours
          const transfers_24h = await this.get_token_transfers_24h(coin.contract);

          // Update the database with new data
          const updated_data = {
            ...coin,
            holders: holders || coin.holders,
            transfers_24h: transfers_24h || coin.transfers_24h,
          };

          await Database.upsert_coin(updated_data);
          updated_coins.push(updated_data);

          console.log(`Updated blockchain data for ${coin.name} (${coin.symbol})`);
        } catch (error) {
          console.error(`Error updating blockchain data for coin ${coin.name}:`, error);
        }
      }

      console.log(`Successfully updated blockchain data for ${updated_coins.length} coins`);
      return updated_coins;
    } catch (error) {
      console.error('Error in blockchain data update:', error);
      return updated_coins;
    }
  }

  static async get_token_holder_count(contract_address) {
    // In a full implementation, this would call Etherscan API to get holder count
    // For now, we'll simulate it with a placeholder
    if (!this.etherscan_api_key) {
      console.warn('ETHERSCAN_API_KEY not set - using mock data for holders');
      return null;
    }

    try {
      const url = `https://api.etherscan.io/api?module=token&action=tokenholderlist&contractaddress=${contract_address}&page=1&offset=1&apikey=${this.etherscan_api_key}`;

      const response = await fetch(url);
      if (!response.ok) {
        throw new Error(`Failed to fetch holder count: ${response.status} ${response.statusText}`);
      }

      const data = await response.json();

      // Extract holder count from response (specific to Etherscan API)
      // Note: This is a simplified example, actual implementation may differ
      if (data.status === '1' && data.message === 'OK') {
        // This is just a sample - Etherscan doesn't directly provide holder count this way
        return parseInt(data.result.length) || null;
      }

      return null;
    } catch (error) {
      console.error('Error fetching token holder count:', error);
      return null;
    }
  }

  static async get_token_transfers_24h(contract_address) {
    // In a full implementation, this would call Etherscan API to get transfer count
    // For now, we'll simulate it with a placeholder
    if (!this.etherscan_api_key) {
      console.warn('ETHERSCAN_API_KEY not set - using mock data for transfers');
      return null;
    }

    try {
      // Calculate timestamp for 24 hours ago
      const timestamp_24h_ago = Math.floor(Date.now() / 1000) - 24 * 60 * 60;

      const url = `https://api.etherscan.io/api?module=account&action=tokentx&contractaddress=${contract_address}&startblock=0&endblock=999999999&sort=asc&apikey=${this.etherscan_api_key}`;

      const response = await fetch(url);
      if (!response.ok) {
        throw new Error(`Failed to fetch transfers: ${response.status} ${response.statusText}`);
      }

      const data = await response.json();

      // Extract transfer count from response (specific to Etherscan API)
      if (data.status === '1' && data.message === 'OK') {
        // Filter transfers in the last 24 hours
        const recent_transfers = data.result.filter(
          (tx) => parseInt(tx.timeStamp) >= timestamp_24h_ago
        );

        return recent_transfers.length;
      }

      return null;
    } catch (error) {
      console.error('Error fetching token transfers:', error);
      return null;
    }
  }
}

module.exports = Blockchain;
